<a href="https://colab.research.google.com/github/Yuliia-Safonova/DTA-2026/blob/main/Homework/Python_practice_05_05_2026.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 🐍 Python: Практикум

**Час виконання:** 3 години  
**Формат:** індивідуальна робота

---

## 📋 Що ви будете практикувати

| Розділ | Тема | Кількість завдань |
|---|---|---|
| 1 | `print`, `input`, змінні | 8 |
| 2 | `if / elif / else` (умови) | 10 |
| 3 | Цикл `for` з `range()` | 10 |
| 4 | Цикл `for` з рядками і списками | 10 |
| 5 | Цикл `while` | 8 |
| 6 | Списки (`list`) — методи та операції | 10 |
| 7 | 🔥 Комбіновані завдання | 6 |

**Усього:** 62 завдання

---

## 📌 Як працювати з зошитом

1. Читайте умову уважно — там може бути більше деталей, ніж здається
2. Не поспішайте — якщо завдання здається складним, почніть з простого варіанту, а потім ускладнюйте
3. Запускайте код часто (Shift + Enter), щоб перевіряти, що все працює
4. Якщо щось не виходить — перечитайте підказку (якщо є), погугліть, поекспериментуйте
5. **Помилки — це нормально!** Кожен програміст бачить їх сотні разів на день

> 💡 У завданнях позначено: ⭐ — легке, ⭐⭐ — середнє, ⭐⭐⭐ — складне

---

# Розділ 1. `print`, `input` та змінні

Базові завдання на введення-виведення та роботу зі змінними.

**Що треба знати:**
- `print(...)` — вивести на екран
- `input("текст: ")` — отримати ввід від користувача (повертає текст!)
- `int(...)`, `float(...)` — перетворити текст у число
- Арифметика: `+`, `-`, `*`, `/`, `**` (степінь), `%` (залишок)

### Завдання 1. Привітання ⭐

Запитайте у користувача його ім'я і виведіть привітання.

**Приклад:**
```
Як вас звати? Олена
Привіт, Олена!
```

In [ ]:
# Ваш код для завдання 1:
name = input("Як Вас звати? ")
print("Привіт,", name + "!")

Як Вас звати? Юлія
Привіт, Юлія!


### Завдання 2. Сума двох чисел ⭐

Запитайте у користувача два числа і виведіть їхню суму.

**Приклад:**
```
Перше число: 7
Друге число: 12
Сума: 19
```

> 💡 **Підказка:** Не забудьте перетворити введення на число через `int(...)`.

In [ ]:
# Ваш код для завдання 2:
x = int(input("Введіть перше ціле число: "))
y = int(input("Введіть друге ціле число: "))
print("Сума чисел:", x + y)

Введіть перше ціле число: 4
Введіть друге ціле число: 7
Сума чисел: 11


### Завдання 3. Площа прямокутника ⭐

Запитайте довжину та ширину прямокутника і виведіть його площу.

In [ ]:
# Ваш код для завдання 3:
a = int(input("Введіть довжину прямокутника (ціле число): "))
b = int(input("Введіть ширину прямокутника (ціле число): "))
print("Площа прямокутника: ", a * b)

Введіть довжину прямокутника (ціле число): 7
Введіть ширину прямокутника (ціле число): 5
Площа прямокутника:  35


### Завдання 4. Сантиметри в метри ⭐

Запитайте у користувача кількість сантиметрів і переведіть їх у метри.

**Приклад:**
```
Кількість см: 250
Це 2.5 м
```

In [ ]:
# Ваш код для завдання 4:
x = int(input("Кількість см: "))
print("Це ", x / 100, "м")

Кількість см: 300
Це  3.0 м


### Завдання 5. Округлення оцінки ⭐⭐

Запитайте у користувача оцінку (число з крапкою, наприклад 4.7).
Виведіть її, округлену до цілого числа.

> 💡 **Підказка:** Використайте функцію `round()`. Не забудьте `float()` замість `int()`.

In [ ]:
# Ваш код для завдання 5:
x = float(input("Назвіть, будь ласка, оцінку (число з крапкою): "))
print("Округлення оцінки: ", round(x))


Назвіть, будь ласка, оцінку (число з крапкою): 9.7
Округлення оцінки:  10


### Завдання 6. Ціна з ПДВ ⭐⭐

Запитайте у користувача ціну товару без ПДВ.
Розрахуйте та виведіть:
- суму ПДВ (20% від ціни)
- кінцеву ціну (ціна + ПДВ)

In [ ]:
# Ваш код для завдання 6:
x = float(input("Введіть ціну товару без ПДВ: "))
print("ПДВ складає:", 0.2 * x)
print("Ціна товару з ПДВ:", 1.2 * x)

Введіть ціну товару без ПДВ: 100
ПДВ складає: 20.0
Ціна товару з ПДВ: 120.0


### Завдання 7. Зворотній відлік ⭐⭐

Запитайте у користувача число `n` і виведіть на одному рядку числа від `n` до 1, розділені пробілами.

**Приклад:**
```
Введіть число: 5
5 4 3 2 1
```

> 💡 **Підказка:** Можна використати `print(n, n-1, n-2, n-3, n-4)` — лише для цього завдання, без циклу.

In [ ]:
# Ваш код для завдання 7:
n=int(input("Введіть перше ціле число: "))
print("Ваш числовий ряд за спаданням: ", *range(n, 0, -1))

Введіть перше ціле число: 7
Ваш числовий ряд за спаданням:  7 6 5 4 3 2 1


### Завдання 8. Обмін значень змінних ⭐⭐⭐

Створіть дві змінні `a = 10` і `b = 20`. Поміняйте їхні значення місцями (щоб `a` стало 20, а `b` — 10) і виведіть результат.

> 💡 **Підказка:** Можна через третю тимчасову змінну, або в Python є дуже короткий спосіб: `a, b = b, a`.

In [ ]:
# Ваш код для завдання 8:
a = float(input("Введіть перше число: "))
b = float(input("Введіть друге число: "))
temp = a
a = b
b = temp
print("Змінемо Ваші числа місцями")
print("Перше число стало: ", a)
print("Друге число стало: ", b)

Введіть перше число: 4.7
Введіть друге число: 5.8
Змінемо Ваші числа місцями
Перше число стало:  5.8
Друге число стало:  4.7


---

# Розділ 2. Умови `if / elif / else`

**Що треба знати:**
- `if умова:` — якщо умова правдива
- `elif умова:` — інакше якщо
- `else:` — у всіх інших випадках
- Оператори порівняння: `==`, `!=`, `>`, `<`, `>=`, `<=`
- Логічні оператори: `and`, `or`, `not`

### Завдання 9. Парне чи непарне ⭐

Запитайте у користувача число і виведіть, парне воно чи непарне.

> 💡 **Підказка:** Парне число ділиться на 2 без остачі: `n % 2 == 0`.

In [ ]:
# Ваш код для завдання 9:
x = int(input("Введіть ціле число: "))
if x % 2 == 0:
    print("Число", x, "- парне")
else:
    print("Число", x, "- непарне")

Введіть ціле число: 7
Число 7 - непарне


### Завдання 10. Додатнє, від'ємне чи нуль ⭐

Запитайте у користувача число і виведіть, додатнє воно, від'ємне чи дорівнює нулю.

In [ ]:
# Ваш код для завдання 10:
x = int(input("Введіть число: "))
if x > 0:
    print ("Число", x, "додатнє")
elif x < 0:
    print("Число", x, "від'ємне")
else:
    print("Число", x, "дорівнює нулю")

Введіть число: -4
Число -4 від'ємне


### Завдання 11. Більше з двох ⭐

Запитайте у користувача два числа і виведіть більше з них.
Якщо числа рівні — виведіть «числа рівні».

In [ ]:
# Ваш код для завдання 11:
x = float(input("Введіть перше число: "))
y = float(input("Введіть друге число: "))
if x > y:
    print("Найбільше з цих чисел: ", x)
elif x < y:
    print("Найбільше з цих чисел: ", y)
else:
    print("Числа рівні")

Введіть перше число: 7
Введіть друге число: 7
Числа рівні


### Завдання 12. Чи може людина голосувати ⭐

Запитайте у користувача вік. Виведіть:
- «Ви можете голосувати» — якщо вік ≥ 18
- «Ви ще малі для голосування» — якщо вік < 18
- «Некоректний вік» — якщо вік від'ємний

In [ ]:
# Ваш код для завдання 12:
age = int(input("Введіть свій вік: "))
if age >= 18:
    print("Ви можете голосувати")
elif age <= 0:
    print("Некоректний вік")
else:
    print("Ви ще малі для голосування")

Введіть свій вік: 0
Некоректний вік


### Завдання 13. Оцінка за бал ⭐⭐

Запитайте у користувача оцінку (від 0 до 100) і виведіть:
- 90+ → «Відмінно»
- 75-89 → «Добре»
- 60-74 → «Задовільно»
- 0-59 → «Не склав»
- інакше (від'ємні або >100) → «Некоректна оцінка»

In [ ]:
# Ваш код для завдання 13:
n = int(input("Введіть свою оцінку: "))
if n > 100:
    print("Некоректна оцінка")
elif n >= 90:
    print("Відмінно")
elif n >= 75:
    print("Добре")
elif n >= 60:
    print("Задовільно")
elif n >= 0:
    print("Не склав")
else:
    print("Некоректна оцінка")


Введіть свою оцінку: 97
Відмінно


### Завдання 14. Найбільше з трьох ⭐⭐

Запитайте у користувача три числа і виведіть найбільше з них.

> 💡 **Підказка:** Можна або послідовно через `if/elif`, або скористатись вбудованою функцією `max()`.

In [ ]:
# Ваш код для завдання 14:
x = float(input("Введіть перше число: "))
y = float(input("Введіть друге число: "))
z = float(input("Введіть третє число: "))
print("Найбільше з цих чисел: ", max(x, y, z))

Введіть перше число: 7
Введіть друге число: 9
Введіть третє число: 4
Найбільше з цих чисел:  9.0


### Завдання 15. Високосний рік ⭐⭐

Запитайте у користувача рік і визначте, чи є він високосним.

**Правила:** Рік є високосним, якщо:
- ділиться на 4, АЛЕ
- НЕ ділиться на 100, АБО
- ділиться на 400

**Приклади:** 2024 — високосний, 2023 — ні, 1900 — ні, 2000 — високосний.

> 💡 **Підказка:** Використайте логічні оператори `and` та `or`. Умова: `(рік % 4 == 0 and рік % 100 != 0) or рік % 400 == 0`.

In [ ]:
# Ваш код для завдання 15:
x = int(input("Введіть рік: "))
if (x % 4 ==0 and x % 100 !=0) or x % 400 == 0:
    print("Цей рік є високосний")
else:
    print("Цей рік не високосний")

Введіть рік: 2007
Цей рік не високосний


### Завдання 16. Пори року ⭐⭐

Запитайте у користувача номер місяця (1-12) і виведіть, до якої пори року він належить:
- 12, 1, 2 → зима
- 3, 4, 5 → весна
- 6, 7, 8 → літо
- 9, 10, 11 → осінь
- інакше → «Некоректний місяць»

In [ ]:
# Ваш код для завдання 16:
month = int(input("Введіть номер місяця (1-12): "))
if month in (12, 1, 2):
    print("Це зима")
elif month in (3, 4, 5):
    print("Це весна")
elif month in (6, 7, 8):
    print("Це літо")
elif month in (9, 10, 11):
    print("Це осінь")
else:
    print("Некоректний місяць")

Введіть номер місяця (1-12): 7
Це літо


### Завдання 17. Калькулятор ⭐⭐

Запитайте у користувача два числа і знак операції (`+`, `-`, `*`, `/`).
Виведіть результат відповідної операції.

**Приклад:**
```
Перше число: 10
Друге число: 3
Знак: *
Результат: 30
```

⚠️ Не забудьте перевірити ділення на 0!

In [ ]:
# Ваш код для завдання 17:
x = float(input("Введіть перше число: "))
y = float(input("Введіть друге число: "))
sign = input("Введіть знак (+, -, *, /): ")

if sign == "+":
    print("Результат операції:", x + y)
elif sign == "-":
    print("Результат операції:", x - y)
elif sign == "*":
    print("Результат операції:", x * y)
elif sign == "/":
    if y == 0:
        print("Помилка: ділити на нуль неможна!")
    else:
        print("Результат операції:", x / y)
else:
    print("Невідомий знак для виконання операції ?!")

Введіть перше число: 7
Введіть друге число: 5
Введіть знак (+, -, *, /): *
Результат операції: 35.0


### Завдання 18. Трикутник ⭐⭐⭐

Запитайте у користувача три довжини сторін і визначте:
1. Чи можна з них скласти трикутник? (правило: сума будь-яких двох сторін має бути більшою за третю)
2. Якщо можна — який це трикутник:
   - рівносторонній (всі сторони рівні)
   - рівнобедрений (дві сторони рівні)
   - різносторонній (всі різні)

> 💡 **Підказка:** Спочатку перевірте, чи трикутник взагалі існує. Потім вже визначайте тип через порівняння сторін.

In [ ]:
# Ваш код для завдання 18:
x = float(input("Введіть першу сторону трикутника: "))
y = float(input("Введіть другу сторону трикутника: "))
z = float(input("Введіть третю сторону трикутника: "))
if (x + y > z) and (y + z > x) and (x + z > y):
    print("Вітаю, такий трикутник існує!")
    if x == y == z:
        print("Це трикутник рівносторонній.")
    elif (x == y) or (x == z) or (y == z):
        print("Цей трикутник рівнобедрений")
    else:
        print("Цей трикутник різносторонній (всі сторони різні)")
else:
    print("Вибачте, але не існує трикутника з такими сторонами.")

Введіть першу сторону трикутника: 3
Введіть другу сторону трикутника: 4
Введіть третю сторону трикутника: 7
Вибачте, але не існує трикутника з такими сторонами.


---

# Розділ 3. Цикл `for` з `range()`

**Що треба знати:**
- `range(n)` → 0, 1, ..., n-1
- `range(a, b)` → a, a+1, ..., b-1
- `range(a, b, step)` → a, a+step, ... (з кроком, не доходячи до b)

### Завдання 19. Числа від 1 до 20 ⭐

Виведіть всі числа від 1 до 20, кожне на новому рядку.

In [ ]:
# Ваш код для завдання 19:
print("Всі числа від 1 до 20:")
for number in range(1, 21):
    print(number)

Всі числа від 1 до 20:
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20


### Завдання 20. Парні числа ⭐

Виведіть всі парні числа від 2 до 30.

> 💡 **Підказка:** Можна через `range(2, 31, 2)` або через перевірку `if n % 2 == 0`.

In [ ]:
# Ваш код для завдання 20:
print("Всі парні числа від 2 до 30:")
for number in range(2, 31, 2):
    print(number)

Всі парні числа від 2 до 30:
2
4
6
8
10
12
14
16
18
20
22
24
26
28
30


### Завдання 21. Сума від 1 до 100 ⭐

Підрахуйте суму всіх чисел від 1 до 100 і виведіть результат. Очікуване число — 5050.

In [ ]:
# Ваш код для завдання 21:
print("Сума всіх чисел від 1 до 100:")
suma = 0
for i in range(1, 101):
    suma = suma + i
print(suma)

Сума всіх чисел від 1 до 100:
5050


### Завдання 22. Таблиця множення ⭐

Запитайте у користувача число і виведіть таблицю множення цього числа на числа від 1 до 10.

**Приклад при введенні 7:**
```
7 * 1 = 7
7 * 2 = 14
...
7 * 10 = 70
```

In [ ]:
# Ваш код для завдання 22:
n = int(input("Введіть число від 1 до 10: "))
print("Таблиця множення для числа", n, ":")
for i in range(1, 11):
    print(n, "*", i, "=", n * i)

Введіть число від 1 до 10: 5
Таблиця множення для числа 5 :
5 * 1 = 5
5 * 2 = 10
5 * 3 = 15
5 * 4 = 20
5 * 5 = 25
5 * 6 = 30
5 * 7 = 35
5 * 8 = 40
5 * 9 = 45
5 * 10 = 50


### Завдання 23. Сума парних чисел ⭐⭐

Підрахуйте суму всіх парних чисел від 1 до 100.

In [ ]:
# Ваш код для завдання 23:
print("Сума всіх парних чисел від 1 до 100:")
suma = 0
for i in range(2, 101, 2):
    suma = suma + i
print(suma)

Сума всіх парних чисел від 1 до 100:
2550


### Завдання 24. Факторіал ⭐⭐

Запитайте у користувача число `n` і обчисліть його факторіал (`n!`).

**Нагадування:** факторіал — це добуток всіх чисел від 1 до `n`.  
Наприклад, `5! = 1 * 2 * 3 * 4 * 5 = 120`.

> 💡 **Підказка:** Створіть змінну `result = 1`, а потім у циклі множте її на кожне число.

In [ ]:
# Ваш код для завдання 24:
n = int(input("Введіть число для обчислення факторіала: "))
result = 1
for i in range(1, n+1):
    result = result * i
print(f"{n}! = {result}")


Введіть число для обчислення факторіала: 5
5! = 120


### Завдання 25. Степінь числа ⭐⭐

Запитайте у користувача два числа: основу `a` і степінь `n`.
Обчисліть `a` в степені `n`, НЕ використовуючи оператор `**`.

**Приклад:** `a = 3, n = 4` → `3 * 3 * 3 * 3 = 81`

> 💡 **Підказка:** Створіть змінну `result = 1` і помножте її на `a` рівно `n` разів.

In [ ]:
# Ваш код для завдання 25:
print("Введіть два числа для обчислення степені.")
a = int(input("Введіть основу (ціле число) a = "))
n = int(input("Введіть степінь (ціле число) n = "))
result = 1
for i in range(n):
    result = result * a
print("Число", a, "в степені", n, "=", result)


Введіть два числа для обчислення степені.
Введіть основу (ціле число) a = 5
Введіть степінь (ціле число) n = 3
Число 5 в степені 3 = 125


### Завдання 26. Сума цифр числа ⭐⭐⭐

Запитайте у користувача число і знайдіть суму його цифр.

**Приклад:** число 1234 → сума 1+2+3+4 = 10

> 💡 **Підказка:** Найпростіший спосіб: перетворити число на рядок через `str()`, а потім перебрати символи у циклі і додавати `int(символ)` до суми.

In [ ]:
# Ваш код для завдання 26:
number = input("Введіть число: ")
suma = 0

for i in number:
    suma = suma + int(i)
print("Сума цифр цього числа:", suma)

Введіть число: 25021978
Сума цифр цього числа: 34


### Завдання 27. Прості числа ⭐⭐⭐

Виведіть всі прості числа від 2 до 50.

**Нагадування:** просте число — це число, яке ділиться тільки на 1 і саме на себе.

> 💡 **Підказка:** Для кожного числа `n` від 2 до 50 перевіряйте, чи має воно дільники від 2 до n-1. Можна використати змінну-прапорець `is_prime = True`.

In [ ]:
# Ваш код для завдання 27:
print("Знайдемо всі прості числа від 2 до 50: ")
for n in range(2, 51):
    is_prime = True
    for d in range(2, n):
        if n % d == 0:
            is_prime = False
            break
    if is_prime:
        print(n, end = " ")

Знайдемо всі прості числа від 2 до 50: 
2 3 5 7 11 13 17 19 23 29 31 37 41 43 47 

### Завдання 28. Малювання трикутника ⭐⭐⭐

Виведіть на екран прямокутний трикутник з зірочок. Висоту запитайте у користувача.

**Приклад при висоті 5:**
```
*
**
***
****
*****
```

> 💡 **Підказка:** Використайте вкладений цикл АБО операцію множення рядка: `'*' * n` повертає рядок з `n` зірочок.

In [ ]:
# Ваш код для завдання 28:
h = int(input("Задайте висоту трикутника (кількість рядків): "))
for i in range(1, h + 1):
    print("* " * i)

Задайте висоту трикутника (кількість рядків): 7
* 
* * 
* * * 
* * * * 
* * * * * 
* * * * * * 
* * * * * * * 


---

# Розділ 4. Цикл `for` з рядками та списками

**Що треба знати:**
- `for letter in "Python":` — перебираємо літери в рядку
- `for item in [1, 2, 3]:` — перебираємо елементи списку
- `len(x)` — довжина рядка або списку
- Перевірка наявності: `if "a" in word:`

### Завдання 29. Довжина слова ⭐

Запитайте у користувача слово і виведіть кількість літер у ньому.

> 💡 **Підказка:** Використайте функцію `len()`.

In [ ]:
# Ваш код для завдання 29:
word = input("Напишіть будь-яке слово: ")
ln = len(word)
print(f"В цьому слові {ln} літер.")

Напишіть будь-яке слово: Python
В цьому слові 6 літер


### Завдання 30. Виведення кожної літери ⭐

Запитайте у користувача слово і виведіть кожну літеру на новому рядку.

In [ ]:
# Ваш код для завдання 30:
word = input("Напишіть будь-яке слово: ")
print("Це слово у стовбчик буде записано так:")
for i in (word):
    print(i)

Напишіть будь-яке слово: Python
Це слово у стовбчик буде записано так:
P
y
t
h
o
n


### Завдання 31. Підрахунок голосних ⭐⭐

Запитайте у користувача слово і підрахуйте кількість голосних літер у ньому.

Голосні: `а е и і о у я ю є ї`.

> 💡 **Підказка:** Створіть лічильник `count = 0`. У циклі перевіряйте: `if letter in "аеиіоуяюєї":` і збільшуйте лічильник.

In [ ]:
# Ваш код для завдання 31:
word = input("Напишіть будь-яке слово: ")
count = 0
for i in word:
    if i in "аеиіоуяюєї":
        count += 1
print(f"В цьому слові {count} голосних літер.")

Напишіть будь-яке слово: МаеиіоуяюєїМ
В цьому слові 10 голосних літер.


### Завдання 32. Перевернуте слово ⭐⭐

Запитайте у користувача слово і виведіть його у зворотному порядку.

**Приклад:** `Python` → `nohtyP`

> 💡 **Підказка:** Створіть порожній рядок `result = ''` і у циклі додавайте кожну літеру НА ПОЧАТОК: `result = letter + result`.

In [ ]:
# Ваш код для завдання 32:
word = input("Напишіть будь-яке слово: ")
print("Ваше слово навпаки виглядає так:", word[::-1])

Напишіть будь-яке слово: World
Ваше слово навпаки виглядає так: dlroW


### Завдання 33. Підрахунок певної літери ⭐⭐

Запитайте у користувача слово і літеру. Виведіть, скільки разів ця літера зустрічається у слові.

In [2]:
# Ваш код для завдання 33:
word = input("Напишіть будь-яке слово: ")
s = input("Яку літеру в цьому слові треба порахувати: ")
count = 0
for i in word:
    if i == s:
        count = count + 1
print(f"В цьому слові буква '{s}' зустрічається {count} разів.")

Напишіть будь-яке слово: молоко
Яку літеру в цьому слові треба порахувати: о
В цьому слові буква 'о' зустрічається 3 разів.


### Завдання 34. Сума елементів списку ⭐

Дано список `numbers = [12, 7, 18, 3, 25, 9, 14]`. Знайдіть і виведіть суму всіх його елементів. Очікуваний результат: 88.

In [ ]:
numbers = [12, 7, 18, 3, 25, 9, 14]
# Ваш код для завдання 34:

count = 0
for i in numbers:
    count = count + i
print(f"Сума елементів списку: {count}")

Сума елементів списку: 88


### Завдання 35. Найбільший і найменший ⭐⭐

Дано список `numbers = [42, 17, 89, 3, 56, 91, 24, 8, 73]`. Знайдіть найбільший і найменший елементи **без використання `max()` та `min()`**.

> 💡 **Підказка:** Створіть змінні `maximum = numbers[0]` і `minimum = numbers[0]`, а потім у циклі порівнюйте.

In [ ]:
numbers = [42, 17, 89, 3, 56, 91, 24, 8, 73]
# Ваш код для завдання 35:

minimum = numbers[0]
maximum = numbers[0]
for i in numbers:
    if i < minimum:
        minimum = i
for j in numbers:
    if j > maximum:
        maximum = j
print("Найменьше число у списку:", minimum)
print("Найбільше число у списку:", maximum)


Найменьше число у списку: 3
Найбільше число у списку: 91


### Завдання 36. Тільки додатні ⭐⭐

Дано список з додатними та від'ємними числами. Виведіть лише додатні.

In [ ]:
numbers = [12, -5, 7, -18, 25, 0, -3, 9, -14, 30]
# Ваш код для завдання 36:

print("Додатні числа цього ряду: ")
for i in numbers:
    if i > 0:
        print(i, end = " ")



Додатні числа цього ряду: 
12 7 25 9 30 

### Завдання 37. Середній бал ⭐⭐

Дано список оцінок студентів. Підрахуйте середній бал.

> 💡 **Підказка:** Середнє = сума / кількість. Кількість можна отримати через `len()`.

In [ ]:
grades = [85, 72, 90, 65, 78, 88, 92, 70, 81, 76]
# Ваш код для завдання 37:

print("Середній бал складає:")
suma = 0
count = 0
for i in grades:
    suma = suma + i
count = len(grades)
x = suma / count
print(x)


Середній бал складає:
79.7


### Завдання 38. Найдовше слово ⭐⭐⭐

Дано список слів. Знайдіть і виведіть найдовше слово.

> 💡 **Підказка:** Створіть змінну `longest = words[0]`, а потім у циклі порівнюйте через `len()`.

In [ ]:
words = ["кіт", "програмування", "Python", "школа", "інформатика", "олівець"]
# Ваш код для завдання 38:

print("Найдовше слово в цьому списку:")
longest = words[0]

for i in words:
    if len(i) > len(longest):
        longest = i
print(longest)


Найдовше слово в цьому списку:
програмування


---

# Розділ 5. Цикл `while`

**Що треба знати:**
- `while умова:` — повторюємо, поки умова правдива
- ⚠️ Завжди має бути щось, що змінює умову — інакше нескінченний цикл!
- Використовуємо `while`, коли наперед не відомо, скільки разів повторювати

### Завдання 39. Лічильник до 10 ⭐

За допомогою циклу `while` виведіть числа від 1 до 10.

In [ ]:
# Ваш код для завдання 39:
i = 0
while i < 10:
    i = i + 1
    print(i, end = " ")

1 2 3 4 5 6 7 8 9 10 

### Завдання 40. Зворотній відлік ⭐

За допомогою циклу `while` виведіть числа від 10 до 1, потім надрукуйте «Старт!».

In [ ]:
# Ваш код для завдання 40:
i = 11
while i > 1:
    i = i - 1
    print(i, end = " ")
print("... Старт!")

10 9 8 7 6 5 4 3 2 1 ... Старт!


### Завдання 41. Подвоєння ⭐⭐

Починаючи з числа 1, подвоюйте його, поки воно менше 1000. Виводьте кожне значення.

**Очікуваний результат:** 1, 2, 4, 8, 16, 32, 64, 128, 256, 512.

In [ ]:
# Ваш код для завдання 41:
i = 1
while i < 1000:
    print(i, end = " ")
    i = i * 2

1 2 4 8 16 32 64 128 256 512 

### Завдання 42. Сума введених чисел ⭐⭐

Просіть користувача вводити числа, доки він не введе 0.
Після цього виведіть суму всіх введених чисел (нуль не враховується).

> 💡 **Підказка:** Створіть змінну `total = 0` і змінну `number = -1` (щоб умова `while number != 0:` спрацювала на початку). Усередині циклу запитуйте число і додавайте до суми.

In [ ]:
# Ваш код для завдання 42:
total = 0
number = -1

while number != 0:
    number = int(input("Введіть число: "))
    if number != 0:
        total = total + number

print(f"Сума всіх чисел: {total}")

Введіть число: -1
Введіть число: 1
Введіть число: 2
Введіть число: -2
Введіть число: 3
Введіть число: 0
Сума всіх чисел: 3


### Завдання 43. Кількість цифр у числі ⭐⭐

Запитайте у користувача додатнє ціле число і підрахуйте, скільки в ньому цифр.

**Приклад:** число 12345 → 5 цифр.

> 💡 **Підказка:** Поки число більше за 0, ділимо його націло на 10 (`n = n // 10`) і збільшуємо лічильник на 1.

In [ ]:
# Ваш код для завдання 43:
n = int(input("Введіть додатнє ціле число: "))
count = 0

while n > 0:
    n = n // 10
    count += 1
print(f"Кількість цифр: {count}")

Введіть додатнє ціле число: 12345
Кількість цифр: 5


### Завдання 44. Перевірка пароля ⭐⭐

Створіть змінну `correct_password = "qwerty123"`.
У циклі `while` запитуйте у користувача пароль, доки він не введе правильний.
Після успішного введення виведіть «Доступ дозволено».

In [ ]:
# Ваш код для завдання 44:
correct_password = "qwerty123"
pw = ""

while pw != correct_password:
    pw = str(input("Введіть пароль: "))
    if pw == correct_password:
        print("Доступ дозволено.")


Введіть пароль: 123
Введіть пароль: йцу
Введіть пароль: qwe123
Введіть пароль: qwerty123
Доступ дозволено.


### Завдання 45. Сума цифр через while ⭐⭐⭐

Запитайте у користувача додатнє ціле число і знайдіть суму його цифр, використовуючи `while`.

**Приклад:** число 1234 → сума 1+2+3+4 = 10.

> 💡 **Підказка:** У циклі: беремо останню цифру через `n % 10`, додаємо до суми, потім відкидаємо її через `n = n // 10`. Цикл працює, поки `n > 0`.

In [ ]:
# Ваш код для завдання 45:
n = int(input("Введіть додатнє ціле число: "))
suma = 0

while n > 0:
    number = n % 10
    suma = suma + number
    n = n // 10

print(f"Сума цифр цього числа: {suma}")

Введіть додатнє ціле число: 12345
Сума цифр цього числа: 15


### Завдання 46. Найбільший спільний дільник ⭐⭐⭐

Запитайте у користувача два додатні числа і знайдіть їхній найбільший спільний дільник (НСД) за алгоритмом Евкліда.

**Алгоритм:** доки `b ≠ 0`: одночасно `a, b = b, a % b`. Коли `b = 0`, відповідь — це `a`.

**Приклад:** НСД(48, 36) = 12.

In [ ]:
# Ваш код для завдання 46:

a = int(input("Введіть перше додатнє число: "))
b = int(input("Введіть друге додатнє число: "))
while b != 0:
    a, b = b, a % b
print("Найбільший спільний дільник цих чисел (НСД) =", a)

Введіть перше додатнє число: 48
Введіть друге додатнє число: 36
Найбільший спільний дільник цих чисел (НСД) = 12


---

# Розділ 6. Списки (`list`) — методи та операції

**Що треба знати:**
- Створення: `[1, 2, 3]`, `[]` (порожній)
- Доступ: `lst[0]`, `lst[-1]` (останній)
- Довжина: `len(lst)`
- Методи: `.append(x)`, `.insert(i, x)`, `.remove(x)`, `.pop()`, `.sort()`, `.reverse()`
- Перевірка: `x in lst`

### Завдання 47. Створення списку ⭐

Створіть список з 5 ваших улюблених фільмів і виведіть його.

In [ ]:
# Ваш код для завдання 47:
print("Мої улюблені фільми за останній час: ")
print()
favorite = ["Диявол носить Prada", "1 + 1", "Кримінальне чтиво", "Джентльмени", "Містер і місіс Сміт"]
print(favorite)

Мої улюблені фільми за останній час: 

['Диявол носить Prada', '1 + 1', 'Кримінальне чтиво', 'Джентльмени', 'Містер і місіс Сміт']


### Завдання 48. Додавання у список ⭐

Створіть порожній список. У циклі `for` додайте до нього числа від 1 до 10. Виведіть результат.

> 💡 **Підказка:** Використайте метод `.append()`.

In [ ]:
# Ваш код для завдання 48:
print("Список чисел від 1 до 10:")
s = []
for i in range(1, 11):
    s.append(i)
print(s)


Список чисел від 1 до 10:
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


### Завдання 49. Перший і останній ⭐

Дано список. Виведіть його перший і останній елементи.

In [ ]:
colors = ["червоний", "жовтий", "зелений", "синій", "фіолетовий"]
# Ваш код для завдання 49:
print(f"Перший елемент цього списку: '{colors[0]}'")
print(f"Останній елемент цього списку: '{colors[-1]}'")

Перший елемент цього списку: 'червоний'
Останній елемент цього списку: 'фіолетовий'


### Завдання 50. Список покупок ⭐⭐

Створіть список покупок з 3-4 товарів. Виконайте такі дії та виводьте список після кожної:
1. Додайте новий товар у кінець
2. Вставте новий товар на початок
3. Видаліть один товар за назвою
4. Замініть товар на іншому місці на щось нове

In [ ]:
# Ваш код для завдання 50:
shopping = ["яйця", "сир", "молоко", "хліб", "вино", "оливки", "хамон"]
print("Список покупок: ", shopping)

shopping.append("масло")
print("Додамо ще один товар у кінець: ", shopping)

shopping.insert(0, "шоколад")
print("Додамо товар на початок списку: ", shopping)

shopping.remove("хліб")
print("Вирішили не брати 'хліб', видаляєсо його із списку: ", shopping)

shopping[3] = "йогурт"
print("Замість молока візьмемо йогурт:   ", shopping)

Список покупок:  ['яйця', 'сир', 'молоко', 'хліб', 'вино', 'оливки', 'хамон']
Додамо ще один товар у кінець:  ['яйця', 'сир', 'молоко', 'хліб', 'вино', 'оливки', 'хамон', 'масло']
Додамо товар на початок списку:  ['шоколад', 'яйця', 'сир', 'молоко', 'хліб', 'вино', 'оливки', 'хамон', 'масло']
Вирішили не брати 'хліб', видаляєсо його із списку:  ['шоколад', 'яйця', 'сир', 'молоко', 'вино', 'оливки', 'хамон', 'масло']
Замість молока візьмемо йогурт:    ['шоколад', 'яйця', 'сир', 'йогурт', 'вино', 'оливки', 'хамон', 'масло']


### Завдання 51. Сортування і реверс ⭐⭐

Дано список чисел. Відсортуйте його за зростанням, виведіть. Потім переверніть, виведіть.

In [ ]:
numbers = [42, 17, 89, 3, 56, 91, 24, 8, 73]
# Ваш код для завдання 51:

numbers.sort()
print("Відсортуємо наш список за зростанням:  ", numbers)

numbers.reverse()
print("А тепер у зворотньому порядку:", numbers)


Відсортуємо наш список за зростанням:   [3, 8, 17, 24, 42, 56, 73, 89, 91]
А тепер у зворотньому порядку: [91, 89, 73, 56, 42, 24, 17, 8, 3]


### Завдання 52. Парні і непарні в окремі списки ⭐⭐

Дано список чисел. Створіть два нові списки: один з парними числами, другий з непарними. Виведіть обидва.

In [ ]:
numbers = [12, 7, 18, 3, 25, 9, 14, 21, 8, 11, 6, 33]
# Ваш код для завдання 52:

even = []
odd = []

for n in numbers:
    if n % 2 == 0:
        even.append(n)
    else:
        odd.append(n)

print("Парні числа з цього списку: ", even)
print("Непарні числа з цього списку: ", odd)


Парні числа з цього списку:  [12, 18, 14, 8, 6]
Непарні числа з цього списку:  [7, 3, 25, 9, 21, 11, 33]


### Завдання 53. Унікальні елементи ⭐⭐

Дано список з повторами. Створіть новий список, у якому кожен елемент зустрічається лише один раз.

> 💡 **Підказка:** Створіть порожній список `unique = []`. Перебирайте оригінальний список і додавайте елемент тільки якщо його ЩЕ НЕМАЄ в `unique`.

In [ ]:
items = ["яблуко", "банан", "яблуко", "груша", "банан", "слива", "яблуко", "груша"]
# Ваш код для завдання 53:

unique = []
for i in items:
    if i not in unique:
        unique.append(i)
print("Початковий список:", items)
print("Унікальний список:", unique)


Початковий список: ['яблуко', 'банан', 'яблуко', 'груша', 'банан', 'слива', 'яблуко', 'груша']
Унікальний список: ['яблуко', 'банан', 'груша', 'слива']


### Завдання 54. Підрахунок входжень ⭐⭐

Дано список і елемент. Підрахуйте, скільки разів цей елемент зустрічається у списку **без використання методу `.count()`**.

In [ ]:
words = ["кіт", "пес", "кіт", "хом'як", "кіт", "пес", "папуга", "кіт"]
target = "кіт"
# Ваш код для завдання 54:

s = 0
for i in words:
    if i == target:
        s = s + 1
print(f"В цьому списку слово '{target}' зустрічається {s} раз(ів).")


В цьому списку слово 'кіт' зустрічається 4 раз(ів).


### Завдання 55. Об'єднання списків ⭐⭐⭐

Дано два списки. Створіть новий список, який містить:
1. Спочатку елементи першого
2. Потім — елементи другого
3. Після — лише ті елементи з обох списків, які зустрічаються в обох

**Приклад:** `[1, 2, 3]` + `[2, 3, 4]` → `[1, 2, 3, 2, 3, 4, 2, 3]` (бо 2 і 3 — спільні).

In [ ]:
list_a = [1, 2, 3, 4, 5]
list_b = [3, 4, 5, 6, 7]
# Ваш код для завдання 55:

list_c = []
for i in list_a:
    if i in list_b:
        list_c.append(i)
print("Загальний список з елементами, що повторюються: ", list_a + list_b + list_c)


Загальний список з елементами, що повторюються:  [1, 2, 3, 4, 5, 3, 4, 5, 6, 7, 3, 4, 5]


### Завдання 56. Сортування без `.sort()` ⭐⭐⭐

Дано список чисел. Відсортуйте його за зростанням **без використання методу `.sort()` та функції `sorted()`**.

> 💡 **Підказка:** Найпростіший алгоритм — «бульбашкове сортування»: у вкладеному циклі порівнюємо сусідні елементи і міняємо їх місцями, якщо ліворуч стоїть більший. Повторюємо, поки не пройдемо весь список без жодної заміни.

In [ ]:
numbers = [42, 17, 89, 3, 56, 91, 24]
# Ваш код для завдання 56:

n = len(numbers)
for i in range(n):
    for j in range(0, n - i - 1):
        if numbers[j] > numbers[j + 1]:
            numbers[j], numbers[j + 1] = numbers[j + 1], numbers[j]

print("Список, відсортований за зростанням:", numbers)


Список, відсортований за зростанням: [3, 17, 24, 42, 56, 89, 91]


---

# Розділ 7. 🔥 Комбіновані завдання

Тут вам знадобиться поєднати кілька різних тем у одному завданні. Це завдання, найбільш близькі до реальних задач, які доводиться розв'язувати програмістам.

### Завдання 57. Меню кафе ⭐⭐

Дано список страв і список їхніх цін (відповідні позиції).
1. Виведіть меню у форматі «Назва — Ціна грн».
2. Підрахуйте і виведіть середню ціну.
3. Виведіть страви, ціна яких вища за середню.

> 💡 **Підказка:** Щоб одночасно перебирати два списки, використовуйте індекс: `for i in range(len(dishes)):`, далі `dishes[i]` і `prices[i]`.

In [ ]:
dishes = ["Борщ", "Вареники", "Салат", "Стейк", "Десерт", "Чай"]
prices = [85, 95, 70, 320, 110, 35]
# Ваш код для завдання 57:

print("* МЕНЮ КАФЕ *")
print()
for i in range(len(dishes)):
    print(f"{dishes[i]} — {prices[i]} грн.")

avg_price = sum(prices) / len(prices)

print()
print(f"Середня ціна страви: {avg_price:.2f} грн.")

print()
print("Страви з ціною, дорожче ніж середня:")
for j in range(len(dishes)):
    if prices[j] > avg_price:
        print(f"{dishes[j]} — {prices[j]} грн.")

* МЕНЮ КАФЕ *

Борщ — 85 грн.
Вареники — 95 грн.
Салат — 70 грн.
Стейк — 320 грн.
Десерт — 110 грн.
Чай — 35 грн.

Середня ціна страви: 119.17 грн.

Страви з ціною, дорожче ніж середня:
Стейк — 320 грн.


### Завдання 58. Шифр Цезаря ⭐⭐⭐

Запитайте у користувача слово (українською або англійською) і число `shift`.
Зашифруйте слово: кожну літеру замініть на ту, що стоїть на `shift` позицій далі в алфавіті.

**Приклад:** слово `abc`, `shift = 2` → `cde`.

> 💡 **Підказка:** Використайте функції `ord(letter)` (повертає код символу) і `chr(num)` (повертає символ за кодом). Для кожної літери: новий код = ord(letter) + shift. Для простоти можна не обробляти переходи через кінець алфавіту.

In [ ]:
# Ваш код для завдання 58:
# Запрашиваем слово и сдвиг
word  = input("Введіть слово (українською або англійською): ")
shift = int(input("Введіть крок для кодування (shift): "))
result = ""
for l in word:
    new_word = ord(l) + shift
    result += chr(new_word)

print(f"Закодоване слово: {result}")

Введіть слово (українською або англійською): Progress
Введіть крок для кодування (shift): 3
Закодоване слово: Surjuhvv


### Завдання 59. Гра «Вгадай число» ⭐⭐⭐

Створіть змінну `secret = 42` (загадане число). У циклі `while` запитуйте у користувача число:
- Якщо введене число менше за загадане — пишіть «Більше»
- Якщо більше — «Менше»
- Якщо вгадав — пишіть «Вітаю!» і завершуйте цикл
- Підрахуйте кількість спроб і виведіть її в кінці

> 💡 **Підказка:** Використайте змінну-прапорець `guessed = False` як умову циклу. Коли користувач вгадав — `guessed = True`.

In [ ]:
# Ваш код для завдання 59:
print("Гра «Вгадай число»")
secret = 42
guessed = False
t = 0

while not guessed:
    num = int(input("Введіть число від 1 до 100: "))
    t += 1

    if num < secret:
        print("Давай більше! ")
    elif num > secret:
        print("Давай менше! ")
    else:
        guessed = True
        print(f"Вітаю! Ви вгадали це число за {t} спроб(и)!")

Гра «Вгадай число»
Введіть число від 1 до 100: 97
Давай менше! 
Введіть число від 1 до 100: 54
Давай менше! 
Введіть число від 1 до 100: 38
Давай більше! 
Введіть число від 1 до 100: 42
Вітаю! Ви вгадали це число за 4 спроб(и)!


### Завдання 60. Аналіз тексту ⭐⭐⭐

Дано рядок з реченням. Виконайте такий аналіз і виведіть результати:
1. Кількість літер (без пробілів)
2. Кількість слів
3. Кількість голосних літер
4. Найдовше слово

> 💡 **Підказка:** Метод `.split()` розбиває рядок на список слів. Наприклад, `'hello world'.split()` повертає `['hello', 'world']`.

In [ ]:
text = "Програмування на Python це цікаво і корисно"
# Ваш код для завдання 60:

text = input("Напишіть своє речення: ")
print()

letters = 0
for el in text:
    if el != " ":
        letters += 1
print(f"Всього літер в цьому реченні (без пробілів): {letters}\n")

words = text.split()
print(f"Всього слів в цьому реченні: {len(words)}\n")

vowels_uk = "аеєиіїоуюяАЕЄИІЇОУЮЯ"
vowels_en = "aeiouAEIOU"
vowels = 0
for v in text:
    if v in vowels_uk or v in vowels_en:
        vowels += 1
print(f"Голосних літер в цьому реченні: {vowels}\n")

longest = ""
for w in words:
    if len(w) > len(longest):
        longest = w
print(f"Найдовше слово в цьому реченні: «{longest}» - {len(longest)} літер.")

Напишіть своє речення: Програмування на Python це цікаво і корисно

Всього літер в цьому реченні (без пробілів): 37

Всього слів в цьому реченні: 7

Голосних літер в цьому реченні: 15

Найдовше слово в цьому реченні: «Програмування» - 13 літер.


### Завдання 61. Магазин: статистика продажів ⭐⭐⭐

Дано список продажів за тиждень (по одному числу на день, починаючи з понеділка).
Виведіть:
1. Загальну виручку за тиждень
2. Середню виручку на день
3. Найкращий день (з найбільшою виручкою) — назвою (Понеділок, Вівторок...)
4. Кількість днів, коли виручка була вищою за середню

> 💡 **Підказка:** Створіть список з назвами днів: `days = ["Понеділок", "Вівторок", ...]`. Індекс найкращого дня знайдіть у циклі (або через `sales.index(max(sales))`).

In [ ]:
sales = [3200, 2800, 4100, 3500, 5200, 7800, 6500]
# Ваш код для завдання 61:

days  = ["Понеділок", "Вівторок", "Середа", "Четвер", "П'ятниця", "Субота", "Неділя"]

total = sum(sales)
print(f"Загальна виручка магазина за тиждень: {total} грн.")

avg_sum = total / len(sales)
print(f"Середня виручка магазина за тиждень: {avg_sum:.2f} грн./день")

best = sales.index(max(sales))
print(f"Найкращий день з продажів: {days[best]} - {sales[best]} грн.")

above_avg = 0
for s in sales:
    if s > avg_sum:
       above_avg += 1
print(f"Днів з виручкою вище середньої: {above_avg} з {len(sales)}")


Загальна виручка магазина за тиждень: 33100 грн.
Середня виручка магазина за тиждень: 4728.57 грн./день
Найкращий день з продажів: Субота - 7800 грн.
Днів з виручкою вище середньої: 3 з 7


### Завдання 62. Турнірна таблиця ⭐⭐⭐

Дано два списки: команди та їхні очки (відповідні позиції).
Виведіть турнірну таблицю, ВІДСОРТОВАНУ за спаданням очок, у такому форматі:
```
1. Шахтар — 67 очок
2. Динамо — 64 очки
...
```

> 💡 **Підказка:** Найпростіший спосіб: у циклі знаходьте команду з максимальними очками, виводьте її, потім видаляйте цю пару (команду і очки) з обох списків. Повторюйте, поки списки не спорожніють.

In [ ]:
teams = ["Динамо", "Шахтар", "Зоря", "Дніпро", "Ворскла", "Олександрія"]
points = [64, 67, 52, 48, 39, 45]
# Ваш код для завдання 62:

teams_copy  = teams.copy()
points_copy = points.copy()

print("* ТУРНІРНА ТАБЛИЦЯ ФУТБОЛЬНИХ КОМАНД *\n")

place = 1
while teams_copy:
    best_idx = points_copy.index(max(points_copy))
    best_team   = teams_copy[best_idx]
    best_points = points_copy[best_idx]

    print(f"{place}. {best_team} — {best_points} очок")

    teams_copy.pop(best_idx)
    points_copy.pop(best_idx)
    place += 1


* ТУРНІРНА ТАБЛИЦЯ ФУТБОЛЬНИХ КОМАНД *

1. Шахтар — 67 очок
2. Динамо — 64 очок
3. Зоря — 52 очок
4. Дніпро — 48 очок
5. Олександрія — 45 очок
6. Ворскла — 39 очок


---

# 🎉 Молодці!

Якщо ви виконали всі завдання — ви великі молодці! Це справді багато практики, і кожне завдання робило вас сильнішим програмістом.

### 💡 Що робити далі

- Якщо щось залишилось незрозумілим — поверніться до того завдання і спробуйте ще раз
- Перепишіть розв'язки складних завдань з пам'яті — це найкращий спосіб закріпити
- Спробуйте придумати свої варіації цих задач і розв'язати їх

**Не зупиняйтесь — програмування вчиться лише через практику! 🐍**
